File: mini_ep2/mini_ep2.c
/* Mini EP2 - Linguagem Rápida
 * Nome: Lara Ayumi Nagamatsu
 * NUSP: 9910568
 * como compilar:  
 *   digite no terminal:
 *   1.  make mini_rapido
 *   isso gerará o executável "mini", rode no terminal:
 *   2.1  ./mini
 *   obs: caso queira testar o tempo, rode no terminal:
 *   2.2  time ./mini
 */

 //reduzir o tamanho do vetor pela metade pulando os números pares, isso permite criar um vetor com a metade do tamanho
 //usar bits para guardar os Booleanos. A ideia é que você use uma estrutura baseada em bit arrays para reduzir mais ainda o tamanhodo vetor

#include <stdio.h>  //printf, scanf
#include <stdlib.h> //malloc, free

int main() {
    int N;
    printf("Insira o N: ");
    scanf("%d", &N);
    N = 1 << N;       // operação de bitshift

    int primos = 0;
    int primosEspeciais = 0;
    int *crivo = malloc((N/2)*sizeof(int));
    int actual = 0;

    for(int i = 0; i <= N/2; i++) {
        crivo[i] = 1;
    }

    // contabilizando 2:
    primos++;
    primosEspeciais++;

    // i representa o número de vdd
    for(int i = 1; i < N/2; i++) {
        if(crivo[i] == 1) {
            primos++;
            actual = (i*2)+1;
            //printf("%d ", actual);
            
            if((actual)%4 != 3)
                primosEspeciais++;
            
            for(int j = actual*3; j < N; j = j + 2*actual)
                crivo[j/2] = 0;
        }
    }

    printf("Primos: %d\nPrimos especiais: %d\n", primos, primosEspeciais);
    
    free(crivo);  // libera memória alocada

    return 0;
}

File: mini_ep2/mini_ep2.py
# Mini EP2 - Linguagem lenta
# Nome: Lara Ayumi Nagamatsu
# NUSP: 9910568
# como rodar: 
#   digite no terminal:
#   1.1 python3 mini_ep2.py
#   obs: caso queira medir o tempo, rode no terminal:
#   1.2 time python3 mini_ep2.py

# implementados: contabilização apenas de ímpares

import numpy as np

def main():
    N = 1 << int(input("Digite o N: "))
    primos = 0
    primosEspeciais = 0
    crivo = [None] * (N//2)

    for i in range(0, N//2):
        crivo[i] = True

    primos = primos + 1
    primosEspeciais = primosEspeciais + 1

    for i in range(1, (N//2)):
        if crivo[i] == True:
            primos = primos + 1
            actual = (i*2) + 1
            
            if (actual)%4 != 3:
                primosEspeciais = primosEspeciais +1

                #retirar o primeiro? começar a contar do 1

            # quero pegar todos os indices multiplos de actual e impares para
            # depois tornar False esses indices
            #aux = np.arange(1, N+1)
            #aux2 = np.remainder(aux, 2)       # pegar indices impares, i.e. aux%2 != 0
            #aux3 = np.remainder(aux, actual)  # pegar indices multiplos de actual i.e. aux%actual == 0
            #new = np.where(np.logical_and((aux2 != 0), (aux3 == 0)))[0]
            #newList = [x//2 for x in new]
            #del newList[0]
            #newList = np.asarray(newList)

            #for x in newList:
            #    crivo[x] = False

            #print("\n\ncrivo: \n", crivo)
            for j in range(actual*3, N, 2*actual):
                crivo[j//2] = False

    print("Primos: ", primos, "\nPrimos especiais: ", primosEspeciais)

main()